### PR TITLE
Editor: Exposed supported formats in Loader.js;

### PR DESCRIPTION
**Pull request by ycw**
Why refactor:

Editor supports two ways of importing files from fs: accept does guard users from selecting unsupported files in the first place for File>Import, however there's no equivalent facility for the 2nd way, drag n drop, so DIY UI feedback is needed for notifying users that no-op as they dropped unsupported files, I prefer showing Import Error: Unsupported file format (xxx) together with editor's supported file formats (same purpose of accept that supported file formats are listed in browser file dialog), my attempt was using https://github.com/mrdoob/three.js/pull/28295 and https://github.com/mrdoob/three.js/pull/28295#discussion_r1590762209, so 'introspection of supported formats' is inevitable, that's why file handlers are re-organized to help introspect in getSupportedFileFormats().

The PR:

- Extracted file handlers into `fileHandlers` object:
  - The handler interface is `(editor, manager, reader, file, ...handlerSpecificOptions): void` 
  - This object is for internal used
  - To support new file format, simply adds a handler like `fileHandlers[ 'myFormat' ] = myFormatHandler`
  - To remove support of a file format, simply deletes it from `fileHandlers`

- Exposed `getSupportedFileFormats()`:
  - This is a static method 
  - The supported file formats are ... auto-derived ... from `fileHandlers` object
  - Returns array of extension names; sorted by `sort()`
  - One of the consumer is MenuBar.File.js: Use `getSupportedFileFormats()` to generate `accepts` list for File>Import input (not included in this PR)
  - Solved https://github.com/mrdoob/three.js/pull/28295#discussion_r1590734928

### CODE REVIEW

**Comment by mrdoob** [comment]
Hmm, doesn't it make the code less readable?

IMHO maintaining the `input.accept` list doesn't seem too bad because the file browser wouldn't let the files be loaded if not updated.

**Comment by ycw** [comment]
> IMHO maintaining the input.accept list doesn't seem too bad

see https://github.com/mrdoob/three.js/pull/28295#discussion_r1590762209

**Comment by ycw** [comment]
Editor supports two ways of importing files from fs: `accepts` does guard users from selecting unsupported files in the first place for **File>Import**, however there's no equivalent facility for the 2nd way, **drag n drop**, so DIY UI feedback is needed for notifying users that no-op as they dropped unsupported files, I prefer showing `Import Error: Unsupported file format (xxx)` with supported formats alongside (the same purpose of `accepts` that supported formats are listed in browser file dialog) in UI like `alert`, my attempt was using [hardcoded supported formats list here](https://github.com/mrdoob/three.js/pull/28295) and [was rejected](https://github.com/mrdoob/three.js/pull/28295#discussion_r1590762209), so 'introspection of supported formats' is inevitable, that's why file handlers are re-organized to help introspect in `getSupportedFileFormats()`.

If hardcoded list is accepted, then refactoring in this PR is not needed ... meaning that file input's `accepts` value and 'supported file formats hint in UI feedback' can both be hardcoded.

**Comment by Mugen87** [comment]
Since there is now more than one use case that uses the file format list I'm okay with adding it.

File: editor/js/Loader.js [hunk 1]
```diff
@@ -64,931 +64,856 @@ function Loader( editor ) {
...
-					const contents = event.target.result;
+		if ( extension in fileHandlers ) {
 
-					const { Rhino3dmLoader } = await import( 'three/addons/loaders/3DMLoader.js' );
+			if ( extension === 'js' || extension === 'json' ) {
 
-					const loader = new Rhino3dmLoader();
-					loader.setLibraryPath( '../examples/jsm/libs/rhino3dm/' );
-					loader.parse( contents, function ( object ) {
+				fileHandlers[ extension ]( editor, manager, reader, file, this.texturePath );
```
Comment (Mugen87) [inline editor/js/Loader.js:83]: To make this a bit more readable, can you write it like so:
```js
const handleRequest = fileHandlers[ extension ];
```
And later in the `if` block:
```js
handleRequest( editor, manager, reader, file );
```

**Comment by Mugen87** [comment]
Do you mind resolving the merge conflicts as well?

File: editor/js/Loader.js (hunk 1 above)
Comment (ycw) [reply editor/js/Loader.js]: ok

**Comment by ycw** [comment]
done

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
As far as I understand, the main purpose of this PR was to enable `Loader.getSupportedFileFormats()`.

Now that #28396 has made the method no longer needed, should we revert this PR?

**Comment by Mugen87** [comment]
TBH, there is no reason anymore for the refactoring. If you prefer the previous code style, I'm fine with reverting, too.

**Comment by mrdoob** [comment]
What code style do you prefer?

**Comment by Mugen87** [comment]
The previous one, tbh. If would implement the module from scratch, I would rather use a switch/case statement than storing the handlers in an object.

**Comment by mrdoob** [comment]
@ycw

We appreciate all the improvements you're doing to the editor.

However, it may be better to share what you're planing to work on before spending a lot of time working on a PR that changes a lot of files or refactors code.

If you do that, we may be able to help you on coming up with simpler solutions and save you time.

**Comment by ycw** [comment]
> We appreciate all the improvements you're doing to the editor.

Thanks. 

> it may be better to share what you're planing to work 

Had been in https://github.com/mrdoob/three.js/pull/28324#issuecomment-2103259803 Its about UI Feedback, 

> that https://github.com/mrdoob/three.js/pull/28396 has made the method no longer needed, should we revert this PR?

This PR was not originally for #28396, instead it was for #28295, (I even added 'TODO: UI Feedback' [in commit](https://github.com/ycw/three.js/blob/c8ebb78e81be7227e83dfc1f1a5f199cee50fe82/editor/js/Loader.js#L96) :D), so no.

**Pull request merged**